### ISSUE TITLE
Code Review

### ISSUE THREAD

**Issue by samilaine** [gitgabbs/mazesolver#2]
Repository cloned at Tuesday, August 23rd 2022 at 1.44PM.

To begin with, I found code easy to read and logically arranged. However, I sometimes would have liked to see Python [PEP8](https://peps.python.org/pep-0008/) recommended two, instead of one, blank likes between functions -- in my opinion this improves readability of the code (though I strongly disagree on several points PEP8 makes). The few comments peppered in the code were quite helpful when it came to deciphering what is actually being done, though I guess most undergraduates are able to read such well organized code without support of comments.

The code works as described (and expected); so, there is very little to provide in addition to nitpicking as a review. However, some of the nitpicking below might be useufl when working in a project with several individuals instead of doing all the work by oneself.

The [user manual](https://github.com/gitgabbs/mazesolver/blob/main/documentation/user_guide.md) could benefit from the use of a dependency manager like [Python Poetry](https://python-poetry.org/). That way the user could be instructed to run `poetry install` to get PyGame installed, which is less than having to first create a virtual environment and then do `pip install pygame` in it. Furthermore, when using in a multi-developer environment, a dependency manager is probably less of a hassle than having to ensure everyone participating have the same versions of same dependencies installed.

The mention about module path in the [user manual](https://github.com/gitgabbs/mazesolver/blob/main/documentation/user_guide.md) appears, at least in my case, redundant: The implementation seemed to work just fine without any tweaking to the module path:
```
(tiralab-2022-HKGjoufm-py3.10) XXXXXX@XXXXXX ~ % pwd
/Users/XXXXXX
(tiralab-2022-HKGjoufm-py3.10) XXXXXX@XXXXXX ~ % python3 mazesolver/main/main.py 
pygame 2.1.2 (SDL 2.0.18, Python 3.10.4)
Hello from the pygame community. https://www.pygame.org/contribute.html
Width: 3
Height: 3
(tiralab-2022-HKGjoufm-py3.10) XXXXXX@XXXXXX ~ % cd mazesolver 
(tiralab-2022-HKGjoufm-py3.10) XXXXXX@XXXXXX mazesolver % python3 main/main.py  
pygame 2.1.2 (SDL 2.0.18, Python 3.10.4)
Hello from the pygame community. https://www.pygame.org/contribute.html
Width: 7
Height: 7
(tiralab-2022-HKGjoufm-py3.10) XXXXXX@XXXXXX mazesolver % 
```

In class `GenerateMaze` (file `main/generate_maze.py`) instead of reading user input directly, perhaps a parameter for class initializator would be a better choice? That way the class could be usable in other projects. For example like this:
```
    def __init__(self, width, height):
        assert width > 2, "The minimum width of the maze is 3"
        assert height > 3, "The minimum height of the maze is 3"

        self.width = width
        self.height = height
        self.maze = []

        self.init_empty_maze()
```
Another way to approach this would be to [inject a dependency](https://en.wikipedia.org/wiki/Dependency_injection), which would then be used to query (or receive) the width and height of the maze to be generated, for example:
```
# Some other class.
def input_skeleton(prompt, convert, test, error_message):
    while (True):
        try:
            value = convert(input(prompt))
        except ValueError:
           print('Invalid value.')
           continue

        if test(value):
             return value
        else:
             print(error_message)

# A driver class.
reader = lambda: input_skeleton('Width: ',
                                int,
                                lambda x: x > 2,
                                'Minimum is 3')

# MazeGenerator class.
def __init__(inputter):
    width = inputter()
```
This, of course, produces convoluted code, and the example above is overly complex to drive the point home, but this approach is in accord with all the fine principles taught in the introductory programming methodology course ("ohjelmointitekniikka" in Finnish, I presume).

Additionally, in the same `GenerateMaze` class the method for printing the maze on a terminal could be implemented as `__str__()` method, which would then allow an instance of GenerateMaze() to be fed directly to `print()` function instead of having to call its method `print_maze()` explicitly. See, for example [this post in the Medium](https://medium.com/swlh/understanding-repr-and-str-in-python-65dd41538943).

Furthermore, in `GenerateMaze` the method `entrance_exit()` uses a list of indices which is not actually needed. The method could be written as
```
    def entrance_exit(self):
        # creates a random entrance to the maze (1st row)
        while True:
            i = random.randint(0,self.width-1)
            if self.maze[1][i] == ".":
                self.maze[0][i] = "." 
                self.entrance = i 
                break

        while True:
            i = random.randint(0,self.width-1)
            if self.maze[self.height-2][i] == ".":
                self.maze[self.height-1][i] = "." 
                self.exit = i 
                break
```
and a further improvement might be to combine the two `while`-loops, though this is left as an exercise to the reader for laziness reasons.

In the class `FindSolution` (file: `main/find_solution.py`) the algorithm the function names were nicely descriptive. A pedant would at this point stop and mention that in addition to descriptive function names there should also be Python docstrings for each function, explaining a) the arguments a function takes, b) which exceptions, if any, it raises, and
 c) what results might be expected. In the context of this course it seems the lack of docstrings might result a reduced grade; so, it is probably a good move to type in docstrings in `main/find_solution.py` as well as in `main/generate_maze.py`.

In summary, the code was readable, and the implementation worked as described. However, adding docstrings and PEP8-style two blank line separators between each function (method) would improve readability, and with a few minor tweaks code could be more concise (in practically insignificant amounts).
